File: linear-c/src/methods/dd/scattering_source.c
/*
    Minha fome vem da alma;
    Meu machado é de Assis.

    SANDER, J. C. A. - 2026
*/

#include "../../../include/core/common.h"

void
calc_scattering_source(
    const int N,
    const int TOTAL_NODES,
    const double W[SCR N],
    const double H_S0[SCR TOTAL_NODES],
    const double psim[SCR (TOTAL_NODES * N)],
    double ss[SCR TOTAL_NODES])
{
    for (int _ = 0; _ < TOTAL_NODES; _++) {ss[_] = 0.0;}

    const double* restrict actual_psim = &psim[0];

    for (int m = 0; m < N; m++)
    {
        const double w = W[m];
        for (int j = 0; j < TOTAL_NODES; j++)
        {
            ss[j] += w * actual_psim[j];
        }
        actual_psim += TOTAL_NODES;
    }

    for (int _ = 0; _ < TOTAL_NODES; _++) {ss[_] *= H_S0[_];}
}

File: linear-rust/src/methods/dd/scattering_source.rs
/*
    Minha fome vem da alma;
    Meu machado é de Assis.

    SANDER, J. C. A. - 2026
*/

fn calc_scattering_source(n: usize, w: &[f64], h_sigma_s0: &[f64], psim: &[f64], ss: &mut [f64]) {
    ss.iter_mut()
        .zip(psim.chunks_exact(n))
        .zip(h_sigma_s0.iter())
        .for_each(|((ss_j, psim_j), &h_s0_j)| {
            let sum: f64 = w
                .iter()
                .zip(psim_j)
                .map(|(&w_n, &psim_j_n)| w_n * psim_j_n)
                .sum();

            *ss_j = h_s0_j * sum;
        });
}
